irohad/torii/processor/impl/query_processor_impl.cpp

```diff
@@ -22,15 +22,61 @@
 namespace iroha {
   namespace torii {
 
+    bool signaturesSubset(
+        const shared_model::interface::SignatureSetType &signatures,
+        const std::vector<shared_model::crypto::PublicKey> &public_keys) {
+      // TODO 09/10/17 Lebedev: simplify the subset verification IR-510
+      // #goodfirstissue
+      std::unordered_set<std::string> txPubkeys;
+      for (auto sign : signatures) {
+        txPubkeys.insert(sign->publicKey().toString());
+      }
+      return std::all_of(public_keys.begin(),
+                         public_keys.end(),
+                         [&txPubkeys](const auto &public_key) {
+                           return txPubkeys.find(public_key.toString())
+                               != txPubkeys.end();
+                         });
+    }
+
+    std::shared_ptr<shared_model::interface::QueryResponse> buildStatefulError(
+        const shared_model::interface::types::HashType &hash) {
+      return clone(
+          shared_model::proto::TemplateQueryResponseBuilder<>()
+              .queryHash(hash)
+              .errorQueryResponse<
+                  shared_model::interface::StatefulFailedErrorResponse>()
+              .build());
+    }
+
     QueryProcessorImpl::QueryProcessorImpl(
         std::shared_ptr<ametsuchi::Storage> storage)
         : storage_(storage) {}
 
     void QueryProcessorImpl::queryHandle(
         std::shared_ptr<shared_model::interface::Query> qry) {
-      auto qpf = model::QueryProcessingFactory(storage_->getWsvQuery(),
-                                               storage_->getBlockQuery());
-      auto qpf_response = qpf_->execute(*qry);
+      const auto &sig = *qry->signatures().begin();
+
+      const auto &wsv_query = storage_->getWsvQuery();
+      auto qpf =
+          model::QueryProcessingFactory(wsv_query, storage_->getBlockQuery());
+      auto signatories = wsv_query->getSignatories(qry->creatorAccountId());
+      if (not signatories) {
+        subject_.get_subscriber().on_next(buildStatefulError(qry->hash()));
+        return;
+      }
+      bool result = signaturesSubset({sig}, *signatories);
+      if (not result) {
+        auto response =
+            shared_model::proto::TemplateQueryResponseBuilder<>()
+                .queryHash(qry->hash())
+                .errorQueryResponse<
+                    shared_model::interface::StatefulFailedErrorResponse>()
+                .build();
+        subject_.get_subscriber().on_next(clone(response));
+        return;
```

Comment:
Also, first and send branches with return looks similar, maybe combine general code instead?